[ASPA-34] Add endpoint for admin authentication

victor-gan opened the pull request:
**Issue:**
Need to check for authentication when accessing restricted pages

**Solution:**
authenticate() sets cookies with a JWT stored inside if urlkey matches Passkey. If user has cookies with the correct JWT, checkCookie() returns true.

**Risk:**
-

**Reviewed By:**
Joe, Raymond

aprilchen64 committed:
[ASPA-9] Remove name field and get UPI from registration form

jameszu committed:
[ASPA-009]Remove name

victor-gan committed:
add function auth()

aprilchen64 committed:
clean up code

jameszu committed:
Fix bugs about the email part

jameszu committed:
Merge branch 'ASPA-9/Remove-Name-Field--Add-User-ID' of https://github.com/UoaWDCC/ASPA-EnrollmentForm into ASPA-9/Remove-Name-Field--Add-User-ID

aprilchen64 committed:
Merge branch 'master' into ASPA-9/Remove-Name-Field--Add-User-ID

aprilchen64 committed:
fix bugs

jameszu committed:
Final code clean up

victor-gan committed:
Add key matching

joewuthrich committed:
changed console.log to echo

aprilchen64 committed:
ASPA-37 Make admin files

joewuthrich committed:
Added authentication function to give cookie

joewuthrich committed:
Added checking cookie capability

aprilchen64 committed:
[ASPA-37] Add Admin UI styles

victor-gan committed:
Add "iat" to payload

victor-gan committed:
Add link to key stored in auth_props.jsons

The passkey which urlkey has to match is stored in auth_props.json
Added success/failure messages for setting urlkey matching and cookie being set

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+     */
+    public function authenticate() {
+
+        //gets the passkey from private_keys/auth_props.json
+        if (!defined('Passkey')) {
+            $json = file_get_contents("private_keys/auth_props.json");
+            $authProps = json_decode($json, true);
+        
+            define('Passkey', $authProps["auth.passkey"]);
+        }
+        // Scrambler key
+        $jwtKey = 'JWTKEY';
```
A few things with this:
- constants should always be defined as uppercase letters, so change `Passkey` to be like `ADMIN_AUTH_PASSKEY`
- We can move this whole block into `config/constants.php`
- We should also set the scrambler key in `auth_props.json` instead of in the code itself since this is also a very important secret we have to store securely. This can be defined in the same way as Passkey, so maybe a name like `ADMIN_AUTH_PRIVATE_KEY`

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+            return false;
+        }
+
+        $cookieName = 'admin_authentication';
```
Rename the cookie to be specific around ASPA, and move this to the outside of this function (just inside the admin class), like this:

`const AUTH_COOKIE_NAME = "aspa_admin_authentication";`

to access this, you can use `self::AUTH_COOKIE_NAME`

This way, we don't need to redefine it in the `checkCookie()` method

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+     */
+    public function checkCookie() {
+
+        if (!defined('Passkey')) {
+            $json = file_get_contents("private_keys/auth_props.json");
+            $authProps = json_decode($json, true);
+        
+            define('Passkey', $authProps["auth.passkey"]);
+        }
```
Remove this since it's been set once already in `constants.php` (from earlier comment)

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+            "iat" => microtime(),
+        );
+
+        $jwt = JWT::encode($payload, $jwtKey);
```
This line will then be changed to `$jwt = JWT::encode($payload, ADMIN_AUTH_PRIVATE_KEY);`

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+
+        $jwt = JWT::encode($payload, $jwtKey);
+
+        setcookie($cookieName , $jwt); 
```
This can be changed to `setcookie(self::AUTH_COOKIE_NAME, $jwt);`

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+            define('Passkey', $authProps["auth.passkey"]);
+        }
+
+        $jwtKey = 'JWTKEY';
+        $cookieName = 'admin_authentication';
```
Remove these since they exist already (`ADMIN_AUTH_PRIVATE_KEY` and `self::AUTH_COOKIE_NAME`)

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+        $jwtKey = 'JWTKEY';
+        $cookieName = 'admin_authentication';
+
+        if(!isset($_COOKIE[$cookieName])) {
```
Use the new cookie name variable

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+            return false;
+        }
+
+        $jwt = $_COOKIE[$cookieName];
```
Use the new cookie name variable

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+        $jwt = $_COOKIE[$cookieName];
+
+        try {
+            $decoded = JWT::decode($jwt, $jwtKey, array('HS256'));
```
Use new name for `jwtKey`

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+        }
+
+        // needs to be changed
+        if ($decoded->key == Passkey) {
```
Use new name for `Passkey`

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,87 @@ public function markAsPaid() {
+    /**
+     * Check if a user has a specific cookie, and if they do allow them to do different things.
+     */
+    public function checkCookie() {
```
Define this as a private method, since this will likely be used in other methods (inside this admin class). Also, remove all the `echo` in this method since this will be called by internal methods. Maybe do the echo's last since they're still useful for testing purposes?

Ray-F reviewed (changes requested):
Yo @VictorGan2 and @joewuthrich, great work on implementing both the setter and the checker!

I just left some comments on this. Take a look and see if you have any questions. Mostly moving the constant definitions to the `constants.php` file and the refactoring needed to go along with that.

Nice work on implementing the guard clauses as well, genuinely makes code reviews easier! 😄

victor-gan committed:
Move constant declarations to constants.php

Change Passkey to ADMIN_AUTH_PASSKEY

Add ADMIN_AUTH_JWTKEY to auth_props.json

Move constant declarations to constant.php

Ray-F committed:
[ASPA-37] Add page changing structure

aprilchen64 committed:
[ASPA-37] Update elements and add message screen

Ray-F committed:
Merged master into ASPA-37/admin-checkin

Ray-F committed:
[ASPA-37] Add communication with paymentStatus check

mouyang2001 committed:
[ASPA-38] Implement a QR code scanner on the front end (#113)

* [ASPA-38] add qr scanner view and basic scanning functionality

* [ASPA-38] add styling to qr scanner view and a back button

* [ASPA-38] add extra stuff

* [ASPA-38] Integrate QR code scanner into admin checkin app

* [ASPA-38] Change QR code scanner engine

* [ASPA-38] Fix page size on mobile

* [ASPA-38] More fixes to page height on mobile

* [ASPA-38] Fix capitalisation for server environment

Co-authored-by: Raymond Feng <rf.raymondfeng@gmail.com>

Ray-F committed:
Merged ASPA-37/admin-checkin with local changes

aprilchen64 committed:
[ASPA-37] display  messages depending on status Code

aprilchen64 committed:
[ASPA-37] Display messages depending on status code

Ray-F committed:
[ASPA-14] Fix implementation of ASPA-14

Ray-F committed:
[ASPA-37] Make logic for showing different message pages

Ray-F committed:
[ASPA-37] Make check mark users as present

Ray-F committed:
Merge branch 'ASPA-37/admin-checkin' of github.com:UoaWDCC/ASPA-EnrollmentForm into ASPA-37/admin-checkin

Ray-F committed:
[ASPA-37] Remove duplicated code

Ray-F committed:
[ASPA-37] Mark member as present if marking as paid

Ray-F committed:
[ASPA-37] Fix bug with attendance cell

Ray-F committed:
[ASPA-37] Add missing semicolon

Ray-F committed:
[ASPA-37] Make styling changes

Ray-F commented on application/controllers/Admin.php:
```diff
@@ -55,6 +56,68 @@ public function markAsPaid() {
+     */
+    public function checkCookie() {
+
+        $cookieName = 'admin_authentication';
```
Remove this since we have `self::AUTH_COOKIE_NAME`

victor-gan committed:
Make cookie name a global constant

joewuthrich committed:
Merge branch 'master' into ASPA-34/Implement-Authentication-Endpoint

joewuthrich committed:
Merged branch #37 and added checkCookie redirect

joewuthrich committed:
Added redirect for non-authenticated users

Ray-F committed:
Merge branch 'master' into ASPA-34/Implement-Authentication-Endpoint

Ray-F committed:
[ASPA-34] Fix index page loading

Ray-F committed:
[ASPA-34] Improve authenticate endpoint
